Fix handling of batched tensors in cat_rows

Balandat opened the pull request:
This fixes the issue discussed in https://github.com/cornellius-gp/gpytorch/pull/1499#discussion_r594720889: Prior to this PR things break if a batched set of additional rows/cols are added to a LazyTensors via `cat_rows`. This is required for batched fantasizing. This PR also adds unit tests to cover this setting.

The unit test for the batched version with a low-rank root currently fails due to what looks like numerical issues.

vishwakftw commented on gpytorch/lazy/lazy_tensor.py:
```diff
@@ -838,7 +839,11 @@ def cat_rows(self, cross_mat, new_mat, generate_roots=True, **root_decomp_kwargs
                 # solve with least squares
-                new_inv_root = torch.lstsq(Q.transpose(-2, -1), R).solution.transpose(-2, -1)
+                if R.ndim > 2:
+                    # temporary hack around PyTorch limitation, this is no good...
```
Batching is supported in PyTorch master. This is not appropriately reflected in the PyTorch docs, but is supported per my reading of ATen internals.

The input shapes should satisfy the following criteria:
1. `A` should be of dimension >= 2
2. `b` should be of dimension >= 1
3. the difference between `A.dim` and `b.dim` should be 0 and 1
4. if the difference between `A.dim` and `b.dim` is 1, then `b` is unsqueezed in the last dimension i.e., `b.unsqueeze(-1)`.
5. let the shapes of `A` and `b` after step 4 be `A_shape` and `b_shape`. Then `A_shape = b_shape`.

Balandat replied:
Ah ok so this is supported in `torch.linalg.lstsq`, but not in `torch.lstsq`.

Balandat replied:
Also, it's not available in 1.8, so we'll need to wait until we can bump pytorch to 1.9 for this.

jacobrgardner commented:
For an R coming from a QR decomposition is there a reason I'd need a pseudoinverse at all? I am clearly missing something from not being quite up to speed on the previous PR @Balandat

Balandat commented:
> For an R coming from a QR decomposition is there a reason I'd need a pseudoinverse at all?

Well we need the pseudoinverse of `new_inv_root` but I guess you're right that computing the pseudoinverse of `R` doesn't make much sense. We can just construct it directly using the QR decomposiiton's components. Let me modify this.

Balandat commented on gpytorch/lazy/lazy_tensor.py:
```diff
@@ -1609,6 +1594,11 @@ def root_decomposition(self, method: Optional[str] = None):
         from .chol_lazy_tensor import CholLazyTensor
         from .root_lazy_tensor import RootLazyTensor
 
+        # # no need to do extra work if we already have a cached cholesky decomposition
+        # if _is_in_cache_ignore_args(self, "cholesky"):
+        #     res = self.cholesky()
+        #     return CholLazyTensor(res)
```
this is sth that I think makes sense. It currently causes some issues though w/ handling `DiagLazyTensors` as proper special cases of `TriangularLazyTensor`. Will have to revisit this some other time.

Balandat commented:
OK so I kind of went down the rabbit hole why working on this. `cat_rows` was pretty hard to parse b/c of overloaded notation, so I refactored the whole thing. Also fixed some potential issues (e.g. we used `TriangularLazyTensor` without actually being sure that the tensor we used was triangular). 

Introduces a `stable_pinverse` util to clean things up a bit that uses `stable_qr` under the hood. As Jake said there was no reason to compute any pseudoinverses of `R` matrices from a QR decomposition.

wjmaddox commented on gpytorch/lazy/lazy_tensor.py:
```diff
@@ -775,73 +775,58 @@ def cat_rows(self, cross_mat, new_mat, generate_roots=True, **root_decomp_kwargs
+        m, n = E.shape[-2:]
+        R = self.root_inv_decomposition().root.evaluate()  # this is fast if L is triangular
+        F = B_ @ R
+        G = lazify(D - F.matmul(F.transpose(-2, -1))).root_decomposition().root.evaluate()
```
Personally, I find `lower_left` and `schur_root` to be a bit more human comprehensible than `F` and `G` respectively. Along with also doing a better job of pointing out what exactly is going on.

Balandat replied:
The main reason of for changing this was to be consistent with the comments and to dedup some double usage of notation for different things.  I can get back to those var names and add more comments if that is desired.

wjmaddox commented on gpytorch/lazy/lazy_tensor.py:
```diff
@@ -775,73 +775,58 @@ def cat_rows(self, cross_mat, new_mat, generate_roots=True, **root_decomp_kwargs
+        if isinstance(E, TriangularLazyTensor) and isinstance(G, TriangularLazyTensor):
+            # make sure these are actually upper triangular
+            if getattr(E, "upper", False) or getattr(G, "upper", False):
+                raise NotImplementedError
```
Can this just be resolved with a transpose? Perhaps something like `new_root = TriangularLazyTensor(new_root, upper=False)`?

Balandat replied:
Well so we assume from the start that we're going to build a block lower triangular matrix. Here we're in the case where `E` or `G` are triangular, but if either of them is upper triangular then there is no way the whole matrix is going to be. Does this make sense? I don't think we currently would ever get into that situation but I figured better safe than sorry here.

wjmaddox replied:
Ahh yeah, good point. The proper thing to do is probably to use the transpose before we even get to this point then, but I think it's probably unlikely to hit it.

Balandat replied:
yeah, we can fix it when we run into it, at least it's breaking loudly

wjmaddox reviewed (commented):
Just a couple of quick comments but otherwise the algebra looks the same to me.

gpleiss commented:
@Balandat lgtm!
